Task 8 solution

maksumov opened the pull request:
# Task 8 Definition

Task defined at https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task8-docker-elastic-beanstalk/task.md

- [x] [Task 8 Prerequisites](https://github.com/maksumov/rs-cart-api/issues/2)

There are also two linked issues in the [another repository](https://github.com/maksumov/rss-shop-be) created:

- [x] [Task 8.1](https://github.com/maksumov/rss-shop-be/issues/8)
- [x] [Task 8.2](https://github.com/maksumov/rss-shop-be/issues/9)

# Solution URLs

1. App FE URL: https://d3sod2gr47frq2.cloudfront.net/
2. App PR: https://github.com/maksumov/nodejs-aws-fe/pull/10
3. maksumov-cart-api HTTP endpoint URL: http://maksumov-cart-api-dev.eu-west-1.elasticbeanstalk.com/
4. maksumov-cart-api HTTPS URL: https://u7dyaoudn4.execute-api.eu-west-1.amazonaws.com/dev/

# Self-check

- [x] Dockerfile is prepared, image is building. Image size is minimised to be [less than 500 MB](https://github.com/maksumov/rss-shop-be/issues/8#issuecomment-740025250).
![image](https://user-images.githubusercontent.com/9920913/101394811-17e76a00-38da-11eb-82a5-2a9ca6e40d6a.png)
- [x] [Dockerfile](https://github.com/maksumov/rs-cart-api/blob/task-8/Dockerfile) is optimized. Files that change more often and commands that depend on them should be included later, files and commands that change less should be at the top.
- [x] Folders are added to [.dockerignore](https://github.com/maksumov/rs-cart-api/blob/task-8/.dockerignore), with explanations. At least 2 big directories should be excluded from build context. [Elastic Beanstalk application is initialized](https://github.com/maksumov/rss-shop-be/issues/9#issuecomment-740209997).
![image](https://user-images.githubusercontent.com/9920913/101419863-fbf7be80-3901-11eb-8dde-79b05b128148.png)
- [x] [Environment is created](https://github.com/maksumov/rss-shop-be/issues/9#issuecomment-740213101) and the app is deployed to the AWS cloud. You must provide a link to your GitHub repo with Cart API service or PR with created Dockerfile and related configurations (**here it is**).
![image](https://user-images.githubusercontent.com/9920913/101419825-e4203a80-3901-11eb-862c-a8ef1d38732c.png)
- [x] FE application is updated with Cart API endpoint. You must provide a [PR with updates in your FE repository](https://github.com/maksumov/nodejs-aws-fe/pull/10) and OPTIONALLY [link to deployed front-end app](https://d3sod2gr47frq2.cloudfront.net/) which makes proper API calls to your [Cart service](https://u7dyaoudn4.execute-api.eu-west-1.amazonaws.com/dev/ping).

All tasks completed. Expected score: 5/5.

maksumov committed:
init: start task 8 solution

maksumov committed:
feat: docker image size less than 500mb

maksumov committed:
feat: docker image less than 100mb

add docker multi-stage build
fix .dockerignore

maksumov committed:
refactor: refactor Dockerfile

add npm script to build docker image
add npm script to docker history
add node user in production image

maksumov committed:
feat: add docker container start & stop

add npm docker:start script to start container in interactive mode
add npm docker:stop script

maksumov committed:
feat: update app

change ping response
add npm deploy script

maksumov committed:
feat: add pm2

add pm2
refactor Dockerfile
refactor .dockerignore
docker image grow up from 97mb up to 119 mb

kirill-ianchuk commented on .dockerignore:
```diff
@@ -0,0 +1,8 @@
+# Remove Git history - it doesn't needed in container
```
What about `dist` folder?

maksumov replied:
I don't understand the question. Do you suggest adding the dist folder here?

maksumov replied:
Thanks for the suggestion. I implemented it

maksumov committed:
refactor: refactor docker image creation steps

remove COPY . . from production stage
add dist to .dockerignore to prevent COPY it to builder stage
rename development stage to builder

maksumov commented on Dockerfile:
```diff
@@ -0,0 +1,40 @@
+RUN npm ci
+
+# Generate dist
+COPY . .
```
It should be worth to change this line on `COPY ./src ./src`

maksumov committed:
refactor: some little specific Dockerfile optimization
